### PR TITLE
Update of README.md 

### PR DESCRIPTION
**Pull request by TobiasSchalau**
For editing the README.md I created the branch "docs/21-readme". All depending issues (#22, #23, #24, #25) should be executed there. 
According to #21 I started to give a structure to the README. Feel free to contact me if you have some suggestions of improvement.

### CODE REVIEW

**Comment by TobiasSchalau** [comment]
Definition of Done for README generally: README has all sections, written understandably, contain all already available content (short project description, section about all requirements)

**Comment by travela** [comment]
#26 is partly solved with the last merge (except that the graphic is backed in and not adaptable)

File: README.md [hunk 1]
```diff
@@ -14,7 +14,8 @@ Developers and anyone else with a PyTorch or TensorFlow neural network model at
 
 - [Python 3.7](https://www.python.org/)
 - [Adversarial Robustness Toolbox (ART) 1.6.1](https://github.com/Trusted-AI/adversarial-robustness-toolbox)
-- [PyTorch 1.8.1](https://pytorch.org/)
+- [PyTorch](https://pytorch.org/)
+- [Tensorflow](https://www.tensorflow.org/)
```
Comment (fraboeni) [inline README.md:18]: What tensorflow-version is supported? Please add it here!

File: README.md [hunk 2]
```diff
@@ -33,6 +33,25 @@ Developers and anyone else with a PyTorch or TensorFlow neural network model at
 ## Attacks
 
 ### Property Inference Attack
+![plot](docs/Property_Interference_Attacks.png)
+
+The Property Inference Attack aims to detect patterns in the parameters of the target model including properties 
+which the model producer has not intended to reveal. Therefore, the adversary requires a 
+meta-classifier which predicts one of these properties he is interested in.
```
Comment (fraboeni) [inline README.md:40]: I would prefer gender-neutral language. You could use plural (attackers/adversaries) and then "their"

File: README.md [hunk 3]
```diff
@@ -33,6 +33,25 @@ Developers and anyone else with a PyTorch or TensorFlow neural network model at
...
 ### Property Inference Attack
+![plot](docs/Property_Interference_Attacks.png)
+
+The Property Inference Attack aims to detect patterns in the parameters of the target model including properties 
+which the model producer has not intended to reveal. Therefore, the adversary requires a 
+meta-classifier which predicts one of these properties he is interested in. 
+
+The first step of the attack is to generate a set containing k datasets for k shadow classifiers which then, are trained 
+to build the training set for the meta classifier. These datasets could be created by sampling from a larger dataset or
+by integrating more data. Furthermore, it is crucial that one half of the shadow classifiers
```
Comment (fraboeni) [inline README.md:44]: It does not become clear what you mean by "integrating more data" - where should it come from? And where should it be integrated to?

File: privacy_evaluator/attacks/attack_interface.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    :type model: ART-Classifier
+    :param training_data: training data of the model.
+    :type test_accuracy: TODO: What Type?
+    :param test_data: test data of the model.
+    :type test_data: TODO: What Type?
+    """
+
+    def __init__(self, model, training_data, test_data):
+        self.model = model
+        self.training_data = training_data
```
Comment (fraboeni) [inline privacy_evaluator/attacks/attack_interface.py:17]: Is the training_data and test_data a tuple (data, labels)?

File: privacy_evaluator/metrics/accuracy.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+def accuracy_difference(train_accuracy, test_accuracy):
```
Comment (fraboeni) [inline privacy_evaluator/metrics/accuracy.py:1]: I would rename here to train_test_accuracy_difference (such that the name of the function already indicates that it is train-test)

File: privacy_evaluator/metrics/accuracy.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    """Calculate the difference between the train and test accuracy.
+
+    The difference is calculated by subtracting the test accuracy from thee train accuracy. The result is a positive
```
Comment (fraboeni) [inline privacy_evaluator/metrics/accuracy.py:4]: Typo: thee

File: privacy_evaluator/metrics/accuracy.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    :param test_accuracy: Accuracy of the test phase.
+    :type test_accuracy: int, float
+
+    :return: The difference between the train and test accuracy.
+    :rtype: int, float
+    """
+    return train_accuracy - test_accuracy
+
+
+def accuracy_ratio(train_accuracy, test_accuracy):
```
Comment (fraboeni) [inline privacy_evaluator/metrics/accuracy.py:19]: also rename to train_to_test_accuracy_ration or something similar, such that it is clear that you divide train by test

File: privacy_evaluator/metrics/accuracy.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    """Calculate the ratio between the train and test accuracy.
+
+    The ratio is calculated by dividing the test accuracy from thee train accuracy. The result is higher than 1 if the
```
Comment (fraboeni) [inline privacy_evaluator/metrics/accuracy.py:22]: Language: please change to "divide the train accuracy by the test accuracy"

File: privacy_evaluator/models/torch/_train_dcti.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        for i, (data, _) in enumerate(loader):
+            data = data.to(device)
+            output = net(data).argmax(dim=1).cpu()
+            prediction[i * batch_size : i * batch_size + len(output) :] = output
+
+    return prediction
+
+
+def main():
+    transform_train = transforms.Compose(
```
Comment (fraboeni) [inline privacy_evaluator/models/torch/_train_dcti.py:52]: Please add a comment here why you perform these operations

**Review by fraboeni — CHANGES_REQUESTED** [review]
Please change the code according to my comments.

File: privacy_evaluator/attacks/attack_interface.py (hunk 4 above)
Comment (travela) [reply privacy_evaluator/attacks/attack_interface.py:17]: This is actually code from team 1 and they fixed it as far as we know.

File: privacy_evaluator/metrics/accuracy.py (hunk 5 above)
Comment (travela) [reply privacy_evaluator/metrics/accuracy.py:1]: This is actually code from team 1 and they fixed it as far as we know.

File: privacy_evaluator/metrics/accuracy.py (hunk 6 above)
Comment (travela) [reply privacy_evaluator/metrics/accuracy.py:4]: This is actually code from team 1 and they fixed it as far as we know.

File: privacy_evaluator/metrics/accuracy.py (hunk 7 above)
Comment (travela) [reply privacy_evaluator/metrics/accuracy.py:19]: This is actually code from team 1 and they fixed it as far as we know.

File: privacy_evaluator/models/torch/_train_dcti.py (hunk 9 above)
Comment (travela) [reply privacy_evaluator/models/torch/_train_dcti.py:52]: This is actually code from team 1 and they fixed it as far as we know.

File: README.md (hunk 3 above)
Comment (travela) [reply README.md]: This part has been updated in commit 9ea9416.

File: README.md (hunk 1 above)
Comment (travela) [reply README.md]: This was changed in commit 9ef4316.

File: README.md (hunk 2 above)
Comment (travela) [reply README.md]: This was addressed in commit 4faf96c.

File: privacy_evaluator/metrics/accuracy.py (hunk 8 above)
Comment (travela) [reply privacy_evaluator/metrics/accuracy.py:22]: This is actually merged code on the main branch from team 1. They deleted this file yesterday in a2c92363fa47ad8e5e5a52176753621a11ebe57b. This should be reflected once our PR is merged.

**Comment by travela** [comment]
All review requests were resolved and the final version was merged into the team2sprint2 branch. The PR can be found at #102.
